### PR TITLE
DCOS-22332 Graphql spec compliance for field values and aliases

### PR DESCRIPTION
**Pull request by Fabs**
This PR ports a group of tests form graphql-js wrapping them into Observable so they are compatible with our implementation. We believe that by making sure the reference implementation tests pass on our implementation is a great first step.

We also refactored part of the code to be easier to understand.

HOW DO I REVIEW?

My suggestion would be to run checkout and run `npm test -- --watch graphql star`. You can write more test cases or modify the current if you can think of a case where it would break. You can also come up with queries by playing on any query explorer (like the github one on notes), and trying to map then to the starwars example.

NOTES:
When testing manually, we are validating our queries structure here: https://developer.github.com/v4/explorer/

The graphql specs:
https://facebook.github.io/graphql/October2016/#sec-Language.Query-Document

**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [x] Did you add new unit tests?

### CODE REVIEW

File: packages/data-service/src/MINIROADMAP [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// "graphqlObservable error: kind ${definition.kind} is not supported"
```
Comment (DanielMSchmidt) [inline packages/data-service/src/MINIROADMAP]: This should not be commited, right?

File: packages/data-service/src/__tests__/reference/starWarsData.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
+// https://github.com/graphql/graphql-js/blob/master/src/__tests__/starWarsData.js
+/**
```
Comment (DanielMSchmidt) [inline packages/data-service/src/__tests__/reference/starWarsData.js]: Nice that you could port them over

File: packages/data-service/src/graphqlObservable.js [hunk 3]
```diff
@@ -44,21 +45,53 @@ function resolveStep(typeMap, definition, context, parent) {
...
+    const resolvedObservable = resolver.resolve(
       parent,
       args,
       context,
       null // that would be the info
     );
 
+    if (!resolvedObservable) {
+      return Observable.throw(
+        new Error("graphqlObservable error: empty resolution")
```
Comment (DanielMSchmidt) [inline packages/data-service/src/graphqlObservable.js]: I wonder if a custom error with that prefix might be nicer here 🤔

File: packages/data-service/src/graphqlObservable.js [hunk 4]
```diff
@@ -44,21 +45,53 @@ function resolveStep(typeMap, definition, context, parent) {
...
+
     return resolvedObservable.map(emittedResults => {
+      if (!emittedResults) {
+        return Observable.throw(
+          new Error("graphqlObservable error: result not emitted")
+        );
+      }
+
+      // Make the concat map appear
+      // const resultStream = !!emittedResults.map ? emittedResults : Observable.of(resultStream);
```
Comment (DanielMSchmidt) [inline packages/data-service/src/graphqlObservable.js]: Forgotten?

**Comment by ahoskins** [comment]
@Fabs Can you clarify why we are having a PR like this adding star wars related stuff? 😏

**Comment by Fabs** [comment]
sure @ahoskins. This PR is work in progress, so I have not yet cared writing a good description.

There is a reference implementation to graphql, graphql-js, and on this implementation, they have a bunch of tests that use this star wars fake data. Here, we are copying their tests and adapting to our implementation to get to a good test coverage, with little effort.

File: packages/data-service/src/graphqlObservable.js [hunk 5]
```diff
@@ -2,100 +2,149 @@ import { Observable } from "rxjs/Observable";
 import "rxjs/add/observable/throw";
 import "rxjs/add/observable/of";
 import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/do";
 import "rxjs/add/operator/map";
+import "rxjs/add/operator/concatMap";
 
-// TODO: refactor this once it has a more comprehensive implementation
-// of the graphql api
+// TODO: add types with typescript
```
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: Is there JIRA for that?

File: packages/data-service/src/graphqlObservable.js [hunk 6]
```diff
@@ -2,100 +2,149 @@ import { Observable } from "rxjs/Observable";
...
+  }
+
+  if (!(resolvedObservable instanceof Observable)) {
+    return throwObservable("resolver does not return an observable");
   }
 
-  return doc.definitions.reduce((acc, definition) => {
-    const resolvedObservable = resolveStep(
-      schema._typeMap,
+  return resolvedObservable.concatMap(emitted => {
```
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: wonder if there's a convention we could pick from Rx code to name `emitted` better.

**Pull request merged**